File: backend/src/kuuna_backend/integrations/s3.py
from __future__ import annotations

from functools import lru_cache

from typing import Any, cast

import boto3  # type: ignore[import-untyped]
from botocore.client import Config  # type: ignore[import-untyped]
from botocore.exceptions import ClientError  # type: ignore[import-untyped]

from kuuna_backend.config.settings import get_settings


@lru_cache
def get_s3_client() -> Any:
    settings = get_settings()

    return boto3.client(
        "s3",
        endpoint_url=settings.s3_endpoint_url,
        aws_access_key_id=settings.s3_access_key,
        aws_secret_access_key=settings.s3_secret_key,
        region_name=settings.s3_region,
        config=Config(signature_version="s3v4", s3={"addressing_style": "path"}),
    )


def ensure_bucket_exists(bucket_name: str) -> None:
    client = get_s3_client()

    try:
        client.head_bucket(Bucket=bucket_name)
        return
    except ClientError:
        pass

    client.create_bucket(Bucket=bucket_name)


def upload_bytes(
    *,
    bucket_name: str,
    object_key: str,
    data: bytes,
    content_type: str | None,
) -> None:
    ensure_bucket_exists(bucket_name)
    client = get_s3_client()

    extra_args: dict[str, str] = {}
    if content_type:
        extra_args["ContentType"] = content_type

    client.put_object(Bucket=bucket_name, Key=object_key, Body=data, **extra_args)


def create_presigned_get_url(*, bucket_name: str, object_key: str, expires_in_seconds: int = 86400) -> str:
    client = get_s3_client()
    url = client.generate_presigned_url(
        ClientMethod="get_object",
        Params={"Bucket": bucket_name, "Key": object_key},
        ExpiresIn=expires_in_seconds,
    )
    return cast(str, url)

File: services/backend-ts/src/integrations/s3.ts
import {
  CreateBucketCommand,
  GetObjectCommand,
  HeadBucketCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

import { getSettings } from "../config.js";

let cachedClient: S3Client | null | undefined;
const ensuredBuckets = new Set<string>();

function getS3Client(): S3Client | null {
  if (cachedClient !== undefined) {
    return cachedClient;
  }

  const settings = getSettings();
  if (!settings.S3_ENDPOINT_URL && (!settings.S3_ACCESS_KEY || !settings.S3_SECRET_KEY)) {
    cachedClient = null;
    return cachedClient;
  }

  cachedClient = new S3Client({
    region: settings.S3_REGION,
    endpoint: settings.S3_ENDPOINT_URL,
    forcePathStyle: Boolean(settings.S3_ENDPOINT_URL),
    credentials:
      settings.S3_ACCESS_KEY && settings.S3_SECRET_KEY
        ? {
            accessKeyId: settings.S3_ACCESS_KEY,
            secretAccessKey: settings.S3_SECRET_KEY,
          }
        : undefined,
  });
  return cachedClient;
}

export function publicUrlFromKey(objectKey: string, bucketName = getSettings().S3_BUCKET): string | null {
  const settings = getSettings();
  if (!settings.S3_PUBLIC_BASE_URL) {
    return null;
  }
  return `${settings.S3_PUBLIC_BASE_URL.replace(/\/$/, "")}/${bucketName}/${objectKey}`;
}

export async function uploadBytes(input: {
  bucketName?: string;
  objectKey: string;
  data: Uint8Array;
  contentType: string;
}): Promise<void> {
  const client = getS3Client();
  if (!client) {
    return;
  }

  const bucketName = input.bucketName ?? getSettings().S3_BUCKET;
  await ensureBucketExists(client, bucketName);
  await client.send(
    new PutObjectCommand({
      Bucket: bucketName,
      Key: input.objectKey,
      Body: input.data,
      ContentType: input.contentType,
    }),
  );
}

async function ensureBucketExists(client: S3Client, bucketName: string): Promise<void> {
  if (ensuredBuckets.has(bucketName)) {
    return;
  }

  try {
    await client.send(new HeadBucketCommand({ Bucket: bucketName }));
  } catch {
    await client.send(new CreateBucketCommand({ Bucket: bucketName }));
  }

  ensuredBuckets.add(bucketName);
}

export async function createPresignedGetUrl(input: {
  bucketName?: string;
  objectKey: string;
  expiresInSeconds?: number;
}): Promise<string | null> {
  const publicUrl = publicUrlFromKey(input.objectKey, input.bucketName);
  if (publicUrl) {
    return publicUrl;
  }

  const client = getS3Client();
  if (!client) {
    return null;
  }

  return getSignedUrl(
    client,
    new GetObjectCommand({
      Bucket: input.bucketName ?? getSettings().S3_BUCKET,
      Key: input.objectKey,
    }),
    { expiresIn: input.expiresInSeconds ?? 7 * 24 * 60 * 60 },
  );
}
